examples/js/plugins/COGParser.js

```diff
@@ -0,0 +1,306 @@
+/* global itowns, THREE */
+
+/**
+ * Select the best overview level (or the final image) to match the
+ * requested extent and pixel width and height.
+ *
+ * @param {Source} source The COGSource
+ * @param {Extent} requestExtent The window extent.
+ * @param {number} requestWidth The pixel width of the window.
+ * @param {number} requestHeight The pixel height of the window.
+ * @returns {Object} The selected zoom level.
+ */
+function selectLevel(source, requestExtent, requestWidth, requestHeight) {
+    // Number of images  = original + overviews if any
+    const cropped = requestExtent.clone().intersect(source.extent);
+    // Dimensions of the requested extent
+    const extentDimension = cropped.planarDimensions();
+
+    const targetResolution = Math.min(
+        extentDimension.x / requestWidth,
+        extentDimension.y / requestHeight,
+    );
+
+    let level;
+
+    // Select the image with the best resolution for our needs
+    for (let index = source.levels.length - 1; index >= 0; index--) {
+        level = source.levels[index];
+        const sourceResolution = Math.min(
+            source.dimensions.x / level.width,
+            source.dimensions.y / level.height,
+        );
+
+        if (targetResolution >= sourceResolution) {
+            break;
+        }
+    }
+
+    return level;
+}
+
+/**
+ * Returns a window in the image's coordinates that matches the requested extent.
+ *
+ * @param {Source} source The COGSource
+ * @param {Extent} extent The window extent.
+ * @param {number[]} resolution The spatial resolution of the window.
+ * @returns {number[]} The window.
+ */
+function makeWindowFromExtent(source, extent, resolution) {
+    const [oX, oY] = source.origin;
+    const [imageResX, imageResY] = resolution;
+
+    const wnd = [
+        Math.round((extent.west - oX) / imageResX),
+        Math.round((extent.north - oY) / imageResY),
+        Math.round((extent.east - oX) / imageResX),
+        Math.round((extent.south - oY) / imageResY),
+    ];
+
+    const xMin = Math.min(wnd[0], wnd[2]);
+    let xMax = Math.max(wnd[0], wnd[2]);
+    const yMin = Math.min(wnd[1], wnd[3]);
+    let yMax = Math.max(wnd[1], wnd[3]);
+
+    // prevent zero-sized requests
+    if (Math.abs(xMax - xMin) === 0) {
+        xMax += 1;
+    }
+    if (Math.abs(yMax - yMin) === 0) {
+        yMax += 1;
+    }
+
+    return [xMin, yMin, xMax, yMax];
+}
+
+/**
+ * Creates a texture from the pixel buffer(s).
+ *
+ * @param {Source} source The COGSource
+ * @param {Object} buffers The buffers (one buffer per band)
```

Comment:
Add implicit type parameters for correct typechecking:
```suggestion
 * @param {THREE.TypedArray | THREE.TypedArray[]} buffers The buffers (one buffer per band)
 * @param {number} buffers.width
 * @param {number} buffers.height
```